core/arch/arm/kernel/secure_partition.c

```diff
@@ -989,10 +1022,181 @@ const struct ts_ops sp_ops __weak __relrodata_unpaged("sp_ops") = {
 	.dump_state = sp_dump_state,
 };
 
+static TEE_Result process_sp_pkg(uint64_t sp_pkg_pa, TEE_UUID *sp_uuid)
+{
+	enum teecore_memtypes mtype = MEM_AREA_RAM_SEC;
+	struct sp_pkg_header *sp_pkg_hdr = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	tee_mm_entry_t *mm = NULL;
+	struct fip_sp *sp = NULL;
+	uint64_t sp_fdt_end = 0;
+	size_t sp_pkg_size = 0;
+	vaddr_t sp_pkg_va = 0;
+	size_t num_pages = 0;
+
+	/* Map only the first page of the SP package to parse the header */
+	if (!tee_pbuf_is_sec(sp_pkg_pa, SMALL_PAGE_SIZE))
+		return TEE_ERROR_GENERIC;
+
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, SMALL_PAGE_SIZE);
+	if (!mm)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sp_pkg_va = tee_mm_get_smem(mm);
+
+	if (core_mmu_map_contiguous_pages(sp_pkg_va, sp_pkg_pa, 1, mtype)) {
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	sp_pkg_hdr = (struct sp_pkg_header *)sp_pkg_va;
+
+	if (sp_pkg_hdr->magic != SP_PKG_HEADER_MAGIC) {
+		EMSG("Invalid SP package magic");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (sp_pkg_hdr->version != SP_PKG_HEADER_VERSION) {
+		EMSG("Invalid SP header version");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (ADD_OVERFLOW(sp_pkg_hdr->img_offset, sp_pkg_hdr->img_size,
+			 &sp_pkg_size)) {
+		EMSG("Invalid SP package size");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (ADD_OVERFLOW(sp_pkg_hdr->pm_offset, sp_pkg_hdr->pm_size,
+			 &sp_fdt_end) || sp_fdt_end > sp_pkg_hdr->img_offset) {
+		EMSG("Invalid SP manifest size");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	core_mmu_unmap_pages(sp_pkg_va, 1);
```

Comment:
@etienne-lms I also changed this line to not use `mm->size`. Since in line 1047 it's hardcoded to map 1 page, it makes sense here to hardcode unmapping 1 page.